File: apps/android/app/src/main/kotlin/com/linonward/app/feature/reading/ArticleService.kt
package com.linonward.app.feature.reading

import com.linonward.app.BuildConfig
import java.net.HttpURLConnection
import java.net.URI
import java.net.URLEncoder
import java.nio.charset.StandardCharsets
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

fun interface ArticleService {
  suspend fun articles(locale: String): List<ReaderArticle>
}

interface ArticleCache {
  suspend fun load(locale: String): List<ReaderArticle>?
  suspend fun save(locale: String, articles: List<ReaderArticle>)
}

class OfflineArticleService(
  private val remote: ArticleService,
  private val cache: ArticleCache,
) : ArticleService {
  override suspend fun articles(locale: String): List<ReaderArticle> = try {
    remote.articles(locale).also { articles ->
      try {
        cache.save(locale, articles)
      } catch (cancelled: CancellationException) {
        throw cancelled
      } catch (_: Exception) {
        // A cache write must not hide fresh content that is ready to display.
      }
    }
  } catch (cancelled: CancellationException) {
    throw cancelled
  } catch (remoteFailure: Exception) {
    try {
      cache.load(locale) ?: throw remoteFailure
    } catch (cancelled: CancellationException) {
      throw cancelled
    } catch (_: Exception) {
      throw remoteFailure
    }
  }
}

class LiveArticleService(
  private val baseUrl: String = BuildConfig.API_BASE_URL,
  private val fetch: suspend (String) -> String = ::fetchJson,
) : ArticleService {
  override suspend fun articles(locale: String): List<ReaderArticle> {
    val primary = request(localeCode(locale))
    if (primary.isNotEmpty() || !locale.lowercase().startsWith("en")) return primary
    return request("zh")
  }

  private suspend fun request(locale: String): List<ReaderArticle> {
    val endpoint = articleEndpoint(baseUrl, locale) ?: throw ArticleServiceException()
    return ArticleResponseDecoder().decode(fetch(endpoint))
  }
}

class ArticleServiceException : Exception()

internal fun articleEndpoint(baseUrl: String, locale: String): String? {
  val origin = runCatching { URI(baseUrl.trim()) }.getOrNull() ?: return null
  if (origin.scheme !in setOf("http", "https") || origin.host.isNullOrBlank()) return null
  if (origin.userInfo != null || origin.query != null || origin.fragment != null) return null
  val root = baseUrl.trim().trimEnd('/')
  val encoded = URLEncoder.encode(localeCode(locale), StandardCharsets.UTF_8.name())
  return "$root/api/content/articles?locale=$encoded"
}

internal fun localeCode(locale: String) = if (locale.lowercase().startsWith("en")) "en" else "zh"

private suspend fun fetchJson(endpoint: String): String = withContext(Dispatchers.IO) {
  val connection = URI(endpoint).toURL().openConnection() as HttpURLConnection
  try {
    connection.requestMethod = "GET"
    connection.setRequestProperty("Accept", "application/json")
    connection.connectTimeout = 10_000
    connection.readTimeout = 15_000
    if (connection.responseCode != HttpURLConnection.HTTP_OK) throw ArticleServiceException()
    connection.inputStream.bufferedReader().use { it.readText() }
  } finally {
    connection.disconnect()
  }
}

internal class ArticleResponseDecoder {
  private val json = Json { ignoreUnknownKeys = true }

  fun decode(raw: String): List<ReaderArticle> =
    json.decodeFromString<ArticlesResponse>(raw).articles.map(PublishedArticle::readerArticle)
}

@Serializable
private data class ArticlesResponse(val articles: List<PublishedArticle>)

@Serializable
private data class PublishedArticle(
  val authorName: String,
  val content: RichTextNode,
  val coverImageUrl: String?,
  val id: String,
  val publishedAt: String?,
  val title: String,
  val updatedAt: String,
) {
  fun readerArticle() = ReaderArticle(
    author = authorName,
    contentHtml = content.html(),
    cover = coverImageUrl?.let { ReaderArticleImage(title, null, it) },
    id = id,
    publishedAt = publishedAt ?: updatedAt,
    readingMinutes = maxOf(1, (content.characterCount() + 399) / 400),
    title = title,
  )
}

@Serializable
private data class RichTextNode(
  val attrs: Attributes? = null,
  val content: List<RichTextNode>? = null,
  val marks: List<Mark>? = null,
  val text: String? = null,
  val type: String,
) {
  fun characterCount(): Int =
    text.orEmpty().count { !it.isWhitespace() } + content.orEmpty().sumOf(RichTextNode::characterCount)

  fun html(): String {
    if (type == "text") {
      return marks.orEmpty().fold(text.orEmpty().escapedHtml()) { value, mark ->
        when (mark.type) {
          "strong" -> "<strong>$value</strong>"
          "em" -> "<em>$value</em>"
          "code" -> "<code>$value</code>"
          "link" -> mark.attrs?.href?.let { "<a href=\"${it.escapedHtml()}\">$value</a>" } ?: value
          else -> value
        }
      }
    }
    val children = content.orEmpty().joinToString("") { it.html() }
    return when (type) {
      "paragraph" -> "<p>$children</p>"
      "heading" -> if (attrs?.level == 1) "<h2>$children</h2>" else "<h3>$children</h3>"
      "blockquote" -> "<blockquote>$children</blockquote>"
      "bullet_list" -> "<ul>$children</ul>"
      "ordered_list" -> "<ol>$children</ol>"
      "list_item" -> "<li>$children</li>"
      "hard_break" -> "<br>"
      else -> children
    }
  }
}

@Serializable
private data class Attributes(val href: String? = null, val level: Int? = null)

@Serializable
private data class Mark(val attrs: Attributes? = null, val type: String)

private fun String.escapedHtml() =
  replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")
    .replace("\"", "&quot;").replace("'", "&#039;")

File: apps/ios/LinOnward/Features/Reading/ArticleService.swift
import Foundation

protocol ArticleService: Sendable {
  func articles(locale: Locale) async throws -> [ReaderArticle]
}

struct ArticleRequestFactory: Sendable {
  private let baseURL: URL

  init?(baseURL: String?) {
    guard
      let baseURL,
      var components = URLComponents(string: baseURL),
      components.scheme == "http" || components.scheme == "https",
      components.host != nil,
      components.user == nil,
      components.password == nil,
      components.query == nil,
      components.fragment == nil
    else { return nil }

    let path = components.path.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
    components.path = path.isEmpty ? "" : "/\(path)"
    guard let url = components.url else { return nil }
    self.baseURL = url
  }

  func articles(locale: Locale) -> URLRequest? {
    articles(localeCode: Self.localeCode(for: locale))
  }

  func fallbackArticles(locale: Locale) -> URLRequest? {
    guard Self.localeCode(for: locale) == "en" else { return nil }
    return articles(localeCode: "zh")
  }

  private func articles(localeCode: String) -> URLRequest? {
    var components = URLComponents(
      url: baseURL.appending(path: "api/content/articles"),
      resolvingAgainstBaseURL: false
    )
    components?.queryItems = [
      URLQueryItem(
        name: "locale",
        value: localeCode
      )
    ]
    guard let url = components?.url else { return nil }
    return URLRequest(url: url)
  }

  private static func localeCode(for locale: Locale) -> String {
    locale.identifier.lowercased().hasPrefix("en") ? "en" : "zh"
  }
}

struct LiveArticleService: ArticleService {
  private let requests: ArticleRequestFactory
  private let session: URLSession

  init(requests: ArticleRequestFactory, session: URLSession = .authentication) {
    self.requests = requests
    self.session = session
  }

  func articles(locale: Locale) async throws -> [ReaderArticle] {
    guard let request = requests.articles(locale: locale) else {
      throw ArticleServiceError.invalidRequest
    }
    let articles = try await fetch(request)
    guard articles.isEmpty, let fallbackRequest = requests.fallbackArticles(locale: locale) else {
      return articles
    }
    return try await fetch(fallbackRequest)
  }

  private func fetch(_ request: URLRequest) async throws -> [ReaderArticle] {
    let (data, response) = try await session.data(for: request)
    guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
      throw ArticleServiceError.unavailable
    }
    return try ArticleResponseDecoder().articles(from: data)
  }
}

enum ArticleServiceError: Error {
  case invalidRequest
  case unavailable
}

struct ArticleResponseDecoder {
  func articles(from data: Data) throws -> [ReaderArticle] {
    let response = try JSONDecoder().decode(ArticlesResponse.self, from: data)
    return response.articles.map(\.readerArticle)
  }
}

private struct ArticlesResponse: Decodable {
  let articles: [PublishedArticle]
}

private struct PublishedArticle: Decodable {
  let authorName: String
  let content: RichTextNode
  let coverImageUrl: String?
  let id: String
  let publishedAt: String?
  let title: String
  let updatedAt: String

  var readerArticle: ReaderArticle {
    ReaderArticle(
      author: authorName,
      contentHtml: content.html,
      cover: coverImageUrl.map {
        ReaderArticleImage(alt: title, caption: nil, url: $0)
      },
      id: id,
      publishedAt: publishedAt ?? updatedAt,
      readingMinutes: max(1, Int(ceil(Double(content.characterCount) / 400))),
      title: title
    )
  }
}

private struct RichTextNode: Decodable {
  struct Attributes: Decodable {
    let href: String?
    let level: Int?
  }

  struct Mark: Decodable {
    let attrs: Attributes?
    let type: String
  }

  let attrs: Attributes?
  let content: [RichTextNode]?
  let marks: [Mark]?
  let text: String?
  let type: String

  var characterCount: Int {
    (text?.filter { !$0.isWhitespace }.count ?? 0) +
      (content ?? []).reduce(0) { $0 + $1.characterCount }
  }

  var html: String {
    if type == "text" {
      return (marks ?? []).reduce(text?.escapedHTML ?? "") { result, mark in
        switch mark.type {
        case "strong": "<strong>\(result)</strong>"
        case "em": "<em>\(result)</em>"
        case "code": "<code>\(result)</code>"
        case "link": mark.attrs?.href.map { "<a href=\"\($0.escapedHTML)\">\(result)</a>" } ?? result
        default: result
        }
      }
    }

    let children = (content ?? []).map(\.html).joined()
    switch type {
    case "paragraph": return "<p>\(children)</p>"
    case "heading": return attrs?.level == 1 ? "<h2>\(children)</h2>" : "<h3>\(children)</h3>"
    case "blockquote": return "<blockquote>\(children)</blockquote>"
    case "bullet_list": return "<ul>\(children)</ul>"
    case "ordered_list": return "<ol>\(children)</ol>"
    case "list_item": return "<li>\(children)</li>"
    case "hard_break": return "<br>"
    default: return children
    }
  }
}

private extension String {
  var escapedHTML: String {
    replacingOccurrences(of: "&", with: "&amp;")
      .replacingOccurrences(of: "<", with: "&lt;")
      .replacingOccurrences(of: ">", with: "&gt;")
      .replacingOccurrences(of: "\"", with: "&quot;")
      .replacingOccurrences(of: "'", with: "&#039;")
  }
}
